cmd/vfkit/main.go

```diff
@@ -264,3 +278,97 @@ func startIgnitionProvisionerServer(ignitionReader io.Reader, ignitionSocketPath
 	log.Debugf("ignition socket: %s", ignitionSocketPath)
 	return srv.Serve(listener)
 }
+
+// it generates a cloud init image by taking the files passed by the user
+// as cloud-init expects files with a specific name (e.g user-data, meta-data) we check the filenames to retrieve the correct info
+// if some file is not passed by the user, an empty file will be copied to the cloud-init ISO to
+// guarantee it to work (user-data and meta-data files are mandatory and both must exists, even if they are empty)
+// if both files are missing it returns an error
+func generateCloudInitImage(files []string) (string, error) {
+	if len(files) == 0 {
+		return "", nil
+	}
+
+	var userData, metaData []byte
+	for _, path := range files {
+		if path == "" {
+			continue
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return "", err
+		}
+		defer file.Close()
+
+		data, err := io.ReadAll(file)
+		if err != nil {
+			return "", err
+		}
+
+		// based on the file name we save its content on the corresponding var
+		filename := filepath.Base(path)
+		switch filename {
+		case "user-data":
+			userData = data
+		case "meta-data":
+			metaData = data
+		}
+	}
+
+	if len(userData) == 0 && len(metaData) == 0 {
+		return "", fmt.Errorf("cloud-init needs user-data and meta-data files to work")
+	}
+
+	return createCloudInitISO(metaData, userData)
+}
+
+// It generates a temp ISO file containing the metaData and userData passed by the user
+// It also register an exit handler to delete the file when vfkit exits
+func createCloudInitISO(metaData, userData []byte) (string, error) {
+	writer, err := iso9660.NewWriter()
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to create writer: %s", err))
+	}
+
+	defer writer.Cleanup()
+
+	userDataReader := bytes.NewReader(userData)
+	err = writer.AddFile(userDataReader, "user-data")
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to add user-data file: %s", err))
+	}
+
+	metaDataReader := bytes.NewReader(metaData)
+	err = writer.AddFile(metaDataReader, "meta-data")
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to add meta-data file: %s", err))
+	}
+
+	tmpDir := os.TempDir()
+	isoFile, err := os.CreateTemp(tmpDir, "vfkit-cloudinit-")
+	if err != nil {
+		return "", errors.Wrap(err, "unable to create temporary cloud-init ISO file")
+	}
+
```

Comment:
I'd move the call to `util.RegisterExitHandler` here to ensure the file is also deleted when an error occurs.